File: solutions/java/log-levels/1/src/main/java/LogLevels.java
import java.util.regex.Matcher;
import java.util.regex.Pattern;

public class LogLevels {

    
    private static final Pattern LOG_PATTERN = Pattern.compile("\\[(?<level>\\w+)\\]:\\s*(?<message>.*)");

    public static String message(String logLine) {
        Matcher matcher = LOG_PATTERN.matcher(logLine);
        if (matcher.find()) {
            // .trim() remove \r\n e espaços extras conforme solicitado
            return matcher.group("message").trim();
        }
        return "";
    }

    public static String logLevel(String logLine) {
        Matcher matcher = LOG_PATTERN.matcher(logLine);
        if (matcher.find()) {
            // Retorna o nível em minúsculas
            return matcher.group("level").toLowerCase();
        }
        return "";
    }

    public static String reformat(String logLine) {
        // Combinamos os métodos anteriores para criar o novo formato
        return String.format("%s (%s)", message(logLine), logLevel(logLine));
    }
}

File: solutions/cpp/log-levels/1/log_levels.cpp
#include <string>

namespace log_line {

std::string message(const std::string& line) {
    // Encontra a posição de ": " que separa o nível da mensagem
    size_t colon_pos = line.find(": ");
    if (colon_pos == std::string::npos) {
        return "";
    }
    
    // Extrai tudo o que vem após o ": "
    std::string msg = line.substr(colon_pos + 2);
    
    // Trim: Remove possíveis espaços em branco e quebras de linha no início e fim
    size_t first = msg.find_first_not_of(" \t\r\n");
    if (first == std::string::npos) {
        return "";
    }
    size_t last = msg.find_last_not_of(" \t\r\n");
    
    return msg.substr(first, (last - first + 1));
}

std::string log_level(const std::string& line) {
    // O nível do log está sempre entre os colchetes '[' e ']'
    size_t start = line.find('[');
    size_t end = line.find(']');
    
    if (start == std::string::npos || end == std::string::npos || end <= start) {
        return "";
    }
    
    // Retorna apenas o texto contido de forma interna
    return line.substr(start + 1, end - start - 1);
}

std::string reformat(const std::string& line) {
    // Combina o resultado dos métodos anteriores
    std::string msg = message(line);
    std::string level = log_level(line);
    
    return msg + " (" + level + ")";
}

} // namespace log_line